core-processor/src/ext.rs

```diff
@@ -2182,4 +2211,94 @@ mod tests {
         // Sending value below ED is also fine
         assert!(msg.is_ok());
     }
+
+    #[test]
+    fn test_unreserve_no_reimbursement() {
+        let costs = ExtCosts {
+            rent: RentCosts {
+                reservation: CostOf::new(10),
+                ..Default::default()
+            },
+            ..Default::default()
+        };
+
+        // Create "pre-reservation".
+        let (id, gas_reservation_map) = {
+            let mut m = BTreeMap::new();
+            let id = ReservationId::generate(MessageId::new([5; 32]), 10);
+
+            m.insert(
+                id,
+                GasReservationSlot {
+                    amount: 1_000_000,
+                    start: 0,
+                    finish: 10,
+                },
+            );
+
+            (id, m)
+        };
+        let mut ext = Ext::new(
+            ProcessorContextBuilder::new()
+                .with_gas(GasCounter::new(u64::MAX))
+                .with_message_context(MessageContextBuilder::new().build())
+                .with_existential_deposit(500)
+                .with_reservations_map(gas_reservation_map)
+                .with_costs(costs.clone())
+                .build(),
+        );
+
+        // Check all the reseravations are in "existing" state.
+        assert!(ext
+            .context
+            .gas_reserver
+            .states()
+            .iter()
+            .all(|(_, state)| matches!(state, GasReservationState::Exists { .. })));
+
+        // Unreserving existing and checking no gas reimbursed.
+        let gas_before = ext.gas_amount();
+        assert!(ext.unreserve_gas(id).is_ok());
+        let gas_after = ext.gas_amount();
+
+        assert_eq!(gas_after.left(), gas_before.left());
+    }
+
+    #[test]
+    fn test_unreserve_with_reimbursement() {
```

Comment:
`CREATE + USE + TRY UNRESERVE` case is required

Reply:
yup, covered